### PR TITLE
Product quantity is now handled by delta modification

### PR DESCRIPTION
**Pull request by jolelievre**
| Questions         | Answers
| ----------------- | -------------------------------------------------------
| Branch?           | develop
| Description?      | This PR changes the behaviour of the quantity field inside product page. Instead of directly modify the product's quantity the merchant instead indicates a delta modification, meaning he says if he wants to increase the stock by ten or decrease it by 5. This results in less chances of messing with the stock in case it changed in the background, because the modification is applied on the current real stock value no matter what.<br><br>To help the merchant understand this modification better the UX was changed you can see the impact on your changes on the quantity total, and you get a history of recent modifications performed by the employees.
| Type?             | improvement
| Category?         | BO
| BC breaks?        | yes
| Deprecations?     | no
| Fixed ticket?     | ~
| How to test?      | Go in page v2, you can edit the quantity by indicating which delta to apply on it. On page submit and refresh you should see your modification in the history (no more than last five modifications displayed).
| Possible impacts? | ~

---
This change is [Reviewable](https://reviewable.io/reviews/prestashop/prestashop/25869)

### Screenshot

Capture d’écran 2021-09-13 à 16 35 47

### Breaking changes

- In those classes `ProductStockProperties` and `UpdateProductStockInformationCommand`, the `delta` field was renamed into `delta_quantity`
- `ProductStockUpdater` service now rely on the `Configuration` service instead of plain values injection

### CODE REVIEW

File: src/Adapter/Product/Stock/QueryHandler/GetEmployeesStockMovementsHandler.php [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+        $movementsData = $this->stockMovementRepository->getLastEmployeeStockMovements(
+            $stockId,
+            $query->getOffset(),
+            $query->getLimit()
+        );
+
+        $movements = [];
+        foreach ($movementsData as $movementDatum) {
+            $movements[] = new EmployeeStockMovement(
+                (int) $movementDatum['id_stock_mvt'],
```
Comment (matks) [inline src/Adapter/Product/Stock/QueryHandler/GetEmployeesStockMovementsHandler.php:73]: Datum = singular of data ? 😄

File: src/Core/Domain/Product/Stock/ValueObject/StockId.php [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+ */
+
+declare(strict_types=1);
+
+namespace PrestaShop\PrestaShop\Core\Domain\Product\Stock\ValueObject;
+
+use PrestaShop\PrestaShop\Core\Domain\Product\Stock\Exception\ProductStockConstraintException;
+
+/**
+ * Stock identity.
```
Comment (matks) [inline src/Core/Domain/Product/Stock/ValueObject/StockId.php]:
```suggestion
 * Stock identifier
```

File: src/PrestaShopBundle/Form/Admin/Type/DeltaQuantityType.php [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+                'label' => $this->trans('Add or subtract items', 'Admin.Global'),
+                'block_prefix' => 'delta_quantity_delta',
+                'row_attr' => [
+                    'class' => 'delta-quantity-delta-container',
+                ],
+            ])
+        ;
+        $builder->get('quantity')->addViewTransformer(new NumberToLocalizedStringTransformer(0, false));
+
+        // Thanks tot his transformer the input data is the quantity value (as integer) and the output is the
```
Comment (matks) [inline src/PrestaShopBundle/Form/Admin/Type/DeltaQuantityType.php]:
```suggestion
        // Thanks to this transformer the input data is the quantity value (as integer) and the output is the
```

File: templates/bundles/PrestaShopBundle/Admin/TwigTemplateForm/prestashop_ui_kit_base.html.twig [hunk 4]
```diff
@@ -1221,3 +1221,34 @@
...
+    {{ form_row(form.delta) }}
+  </div>
+{% endblock %}
+
+{% block delta_quantity_quantity_widget %}
+  {%- set attr = attr|merge({class: (attr.class|default('') ~ ' delta-quantity-quantity')|trim}) -%}
+  <div {{ block('widget_container_attributes') }}>
+    {{- block('form_widget_simple') -}}
+    <span class="initial-quantity">
+      {{ form.vars.value|e('html') }}
```
Comment (matks) [inline templates/bundles/PrestaShopBundle/Admin/TwigTemplateForm/prestashop_ui_kit_base.html.twig]: Isn't it the default value ? `|e('html')`

I think
```
{{ form.vars.value|e('html') }}
```
is equivalent to
```
{{ form.vars.value }}
```
if Twig [autoescape](https://twig.symfony.com/doc/3.x/api.html) is turned on

File: tests/Unit/Core/Form/IdentifiableObject/DataProvider/ProductFormDataProviderTest.php [hunk 5]
```diff
@@ -505,6 +526,19 @@ private function getDatasetsForStock(): array
         $expectedOutputData['stock']['availability']['available_later_label'] = $localizedValues;
         $expectedOutputData['stock']['availability']['available_date'] = '1969-07-20';
 
+        $expectedOutputData['stock']['quantities']['stock_movements'] = [
+            [
+                'date_add' => '2021-05-24 15:24:32',
+                'employee' => 'Paul Atreide',
```
Comment (matks) [inline tests/Unit/Core/Form/IdentifiableObject/DataProvider/ProductFormDataProviderTest.php:532]: Looks like you are planning to watch Dune soon in theaters? 😄

**Review by matks — DISMISSED** [review]
Code is overall OK for me. I added some suggestions for improvements. Great work !

File: src/Adapter/Product/Stock/QueryHandler/GetEmployeesStockMovementsHandler.php (hunk 1 above)
Comment (jolelievre) [reply src/Adapter/Product/Stock/QueryHandler/GetEmployeesStockMovementsHandler.php:73]: Exactly ^^ Data is the plural form in latin

File: templates/bundles/PrestaShopBundle/Admin/TwigTemplateForm/prestashop_ui_kit_base.html.twig (hunk 4 above)
Comment (jolelievre) [reply templates/bundles/PrestaShopBundle/Admin/TwigTemplateForm/prestashop_ui_kit_base.html.twig]: Alright I'll remove it then

File: tests/Unit/Core/Form/IdentifiableObject/DataProvider/ProductFormDataProviderTest.php (hunk 5 above)
Comment (jolelievre) [reply tests/Unit/Core/Form/IdentifiableObject/DataProvider/ProductFormDataProviderTest.php:532]: Maybe 😅 I just wan to finish (re)reading the book first

**Review by matks — DISMISSED** [review]

File: src/Core/Domain/Product/Stock/ValueObject/StockId.php [hunk 6]
```diff
@@ -0,0 +1,75 @@
...
+     */
+    public function getValue()
+    {
+        return $this->stockId;
+    }
+
+    /**
+     * @param int $stockId
+     */
+    private function assertIntegerIsGreaterThanZero($stockId)
```
Comment (Progi1984) [inline src/Core/Domain/Product/Stock/ValueObject/StockId.php]:
```suggestion
    private function assertIntegerIsGreaterThanZero(int $stockId): void
```

File: src/Core/Domain/Product/Stock/ValueObject/StockId.php [hunk 7]
```diff
@@ -0,0 +1,75 @@
...
+    {
+        $this->assertIntegerIsGreaterThanZero($stockId);
+
+        $this->stockId = $stockId;
+    }
+
+    /**
+     * @return int
+     */
+    public function getValue()
```
Comment (Progi1984) [inline src/Core/Domain/Product/Stock/ValueObject/StockId.php]:
```suggestion
    public function getValue(): int
```

File: src/Core/Domain/Product/Stock/ValueObject/StockId.php [hunk 8]
```diff
@@ -0,0 +1,75 @@
...
+     * @var int
+     */
+    private $stockId;
+
+    /**
+     * @param int $stockId
+     *
+     * @throws ProductStockConstraintException
+     */
+    public function __construct($stockId)
```
Comment (Progi1984) [inline src/Core/Domain/Product/Stock/ValueObject/StockId.php]:
```suggestion
    public function __construct(int $stockId)
```

File: src/Core/Domain/Product/Stock/Command/UpdateProductStockInformationCommand.php [hunk 9]
```diff
@@ -136,19 +136,19 @@ public function setPackStockType(int $packStockType): UpdateProductStockInformat
...
     }
 
     /**
-     * @param int $quantity
+     * @param int $deltaQuantity
      *
      * @return UpdateProductStockInformationCommand
      */
-    public function setQuantity(int $quantity): UpdateProductStockInformationCommand
+    public function setDeltaQuantity(int $deltaQuantity): UpdateProductStockInformationCommand
```
Comment (Progi1984) [inline src/Core/Domain/Product/Stock/Command/UpdateProductStockInformationCommand.php:149]: BC Break to document

**Review by Progi1984 — COMMENTED** [review]
@jolelievre Thank you for your contribution. I have brought you some comments

File: src/Core/Domain/Product/Stock/Command/UpdateProductStockInformationCommand.php (hunk 9 above)
Comment (jolelievre) [reply src/Core/Domain/Product/Stock/Command/UpdateProductStockInformationCommand.php:149]: Already done in the PR description

**Review by Progi1984 — DISMISSED** [review]

**Review by matks — DISMISSED** [review]

**Review by Julievrz — DISMISSED** [review]

File: admin-dev/themes/new-theme/scss/pages/product/product_page.scss [hunk 10]
```diff
@@ -1075,6 +1075,37 @@ $product-page-padding-bottom: 80px !default;
...
+      font-size: 0.7rem;
+      color: #fff;
+      text-align: center;
+
+      &.increased_quantity {
+        background-color: $brand-primary;
+      }
+
+      &.decreased_quantity {
+        background-color: #373a40;
```
Comment (NeOMakinG) [inline admin-dev/themes/new-theme/scss/pages/product/product_page.scss]: Could you moove this line to a variable? Or maybe can we choose an existing variable color to avoid inconsistency in our codebase?

File: templates/bundles/PrestaShopBundle/Admin/TwigTemplateForm/entity_search_input.html.twig [hunk 11]
```diff
@@ -48,7 +50,9 @@
       {%- set id = form.vars.id ~ '_search_input' -%}
       <input {{ block('widget_container_attributes') }} />
     </div>
+    {% endif %}
```
Comment (NeOMakinG) [inline templates/bundles/PrestaShopBundle/Admin/TwigTemplateForm/entity_search_input.html.twig]: We don't have a twig linter unfortunatly but I hate when conditions arent aligned as HTML tags, it's way less readable to me...

File: templates/bundles/PrestaShopBundle/Admin/TwigTemplateForm/entity_search_input.html.twig [hunk 12]
```diff
@@ -86,7 +90,11 @@
               <th>{{ child.vars.label }}</th>
             {% endif %}
           {% endfor %}
+
+          {# If item deletion is disabled we don't display the extra column for actions #}
+          {% if form.vars.allow_delete %}
             <th></th>
+          {% endif %}
```
Comment (NeOMakinG) [inline templates/bundles/PrestaShopBundle/Admin/TwigTemplateForm/entity_search_input.html.twig]: Same comment about tabulation, hate this :/

File: templates/bundles/PrestaShopBundle/Admin/TwigTemplateForm/entity_search_input.html.twig [hunk 13]
```diff
@@ -131,8 +139,12 @@
     </td>
     {% endif %}
   {% endfor %}
+
+  {# If item deletion is disabled we don't display the extra column for actions #}
+  {% if form.parent.vars.allow_delete %}
     <td>
       <i class="material-icons delete entity-item-delete">clear</i>
     </td>
+  {% endif %}
```
Comment (NeOMakinG) [inline templates/bundles/PrestaShopBundle/Admin/TwigTemplateForm/entity_search_input.html.twig]: same

edit: my bad 💋

**Comment by NeOMakinG** [comment]
Overall, front-end part looks pretty fine to me, great SCSS stylesheet and nice types (even if there is a JS file that we should migrate to TS, I may did it in another PR I don't remember of), but I left some comments about a variable and some twig codestyle 👍

File: admin-dev/themes/new-theme/scss/pages/product/product_page.scss (hunk 10 above)
Comment (jolelievre) [reply admin-dev/themes/new-theme/scss/pages/product/product_page.scss]: Yep you're right, I was able to find the appropriate variable in the stock scss

File: templates/bundles/PrestaShopBundle/Admin/TwigTemplateForm/entity_search_input.html.twig (hunk 11 above)
Comment (jolelievre) [reply templates/bundles/PrestaShopBundle/Admin/TwigTemplateForm/entity_search_input.html.twig]: We already had this discussion and we don't see things the same way ^^ But I'm not gonna argue with you for this, especially since it seems to cause you physical pain 😅

**Review by NeOMakinG — DISMISSED** [review]

**Review by Progi1984 — DISMISSED** [review]

**Review by NeOMakinG — DISMISSED** [review]

**Review by Progi1984 — DISMISSED** [review]

**Comment by florine2623** [comment]
Hello @jolelievre ,

Thanks for the PR !

I have some feedback to share.

When the delta field is empty, NaN is displayed :
Screenshot 2021-10-01 at 15 11 22

**For Combination products**
I want to subtract 10 items - the arrow with new stock amount isn't displayed :
Screenshot 2021-10-01 at 17 54 18
After saving the edition, delta number "-10" is displayed instead of "50" :
Screenshot 2021-10-01 at 17 54 27
List of latest Stock edition is not updated :
Screenshot 2021-10-01 at 17 59 44
Same case for all type of combinations (default or not).

**History of Stock changes :** 
Stock movement change can be unclear. Only the column Employee is specified, the merchant might not know if Stock change comes from Customer (Order) or Employee (BO).
Maybe a column "Type" can be added like it is in Stock > Movement :
Screenshot 2021-10-01 at 18 10 01
It would also fill up the big space :
Screenshot 2021-10-01 at 18 06 25


**_Miscellaneous - questions :_** if it needs to be created on another issue, let me know !
**For Pack of product**
Is this feature implemented for Product page v2 ?
If not, maybe a message like this can be displayed :
Screenshot 2021-10-01 at 18 03 41

"View all stock movement" from the product page opens a new Stock page, but once you get to that page, the affected product is not specified/displayed anywhere (not on the search bar, nor the filters) - only in the URL :
Screenshot 2021-10-01 at 18 14 54

Thanks!

**Comment by jolelievre** [comment]
Hi @florine2623 

I just fixed the bug you mentioned with `NaN` Regarding your other questions:

> For Combination products
> I want to subtract 10 items - the arrow with new stock amount isn't displayed :

This PR only deals with standard product not combinations, the idea is to handle the combinations in a different PR to avoid making this one too big. I though it would have no impact but I forgot that both forms share the same form types so it impacted the combination form as well, my bad.

I tried to integrate it quickly but I soon realized it represented too much work (besides I found some bugs related to suppliers that need to be handled), so I will follow the initial plan and deal with this in another PR.

So all the bugs you found on combination are to be ignored and will be handled later, sorry if the PR description lacked precision about that.

> History of Stock changes :
> Stock movement change can be unclear. Only the column Employee is specified, the merchant might not know if Stock change comes from Customer (Order) or Employee (BO).

Actually, this component displays the modifications performed by **Employees only** the idea is to only see the modifications made in the BO, not all the ones related to cart or orders. So the type would be irrelevant in this case, however, if you didn't understand it at first sight it probably means that the wording, at least, can be improved.

I followed the mockups for the design and wording, but these are not the definitive design on which Tristan and Julie are still working, the purpose of the PR is to be functional not perfect. So we will improve this later, but thanks for the feedback!

> For Pack of product
> Is this feature implemented for Product page v2 ?

The feature is not implemented yet, I'm waiting for the related product PR to be merged to implement it, we forgot to add a placeholder indeed, it would have been nice for the 178, but at this point, it's not relevant to add it since it should be fully implemented in the coming weeks.

> "View all stock movement" from the product page opens a new Stock page, but once you get to that page, the affected product is not specified/displayed anywhere (not on the search bar, nor the filters) - only in the URL :

This is related to the stock movement page, not the product page. It only offers filtering by product ID in the URL but it doesn't prefill the name or anything. The purpose here was to have a shortcut to the page showing only the edited product. Any new feature will have to be specified by @MatShir and be developed in a dedicated issue/PR.

### TLDR

Thanks for your feedback, they are all relevant but for the scope of this PR only the one regarding `NaN` has been handled. The combination edition will be dealt with later, and for stock history, I made the most out of the info I had. Any additional feature will need a dedicated issue.

**Comment by florine2623** [comment]
Hello @jolelievre ,

Thanks for all the details ! 

I've tested again this PR on only Standard Products, no bugs have been found and `NaN` is fixed !
Before putting QA approved label, is it possible to get a code review ?

> Actually, this component displays the modifications performed by Employees only the idea is to only see the modifications made in the BO, not all the ones related to cart or orders. So the type would be irrelevant in this case, however, if you didn't understand it at first sight it probably means that the wording, at least, can be improved.

It is very clear, I was just wondering if only showing the Employees' modifications was the wanted behavior :)
So, if product page v2 specs are available, I'll be glad to read them !

**Comment by MatShir** [comment]
Thanks for the deep tests and feedback @florine2623. You have a quick eye for detail 
If the `NaN` issue is fixed, there is no blocking issue left @jolelievre 🚀 . I will make an issue for the link and make it possible to filter the page with a product ID and for the different type of product.
Specification are coming 🚧

**Review by Progi1984 — APPROVED** [review]

File: src/Adapter/Product/Stock/Repository/StockMovementRepository.php [hunk 14]
```diff
@@ -0,0 +1,80 @@
...
+    }
+
+    /**
+     * @param StockId $stockId
+     * @param int $offset
+     * @param int $limit
+     *
+     * @return array<array<string, mixed>>
+     */
+    public function getLastEmployeeStockMovements(StockId $stockId, int $offset = 0, int $limit = 3): array
```
Comment (sowbiba) [inline src/Adapter/Product/Stock/Repository/StockMovementRepository.php:65]: why 3? is it necessary to have a constant for it ?

File: src/PrestaShopBundle/Form/Admin/Sell/Product/EventListener/CombinationListener.php [hunk 15]
```diff
@@ -0,0 +1,66 @@
...
+    public function adaptCombinationForm(FormEvent $event): void
+    {
+        $form = $event->getForm();
+        $data = $event->getData();
+
+        if ($form->has('stock')) {
+            $stock = $form->get('stock');
+            if ($stock->has('quantities')) {
+                $quantities = $stock->get('quantities');
+                if ($quantities->has('stock_movements') && empty($data['stock']['quantities']['stock_movements'])) {
```
Comment (sowbiba) [inline src/PrestaShopBundle/Form/Admin/Sell/Product/EventListener/CombinationListener.php:60]: empty($quantities->get('stock_movements'))  ?

**Review by sowbiba — COMMENTED** [review]
just few notes...
YAML linter is still not happy

File: src/PrestaShopBundle/Form/Admin/Sell/Product/EventListener/CombinationListener.php (hunk 15 above)
Comment (jolelievre) [reply src/PrestaShopBundle/Form/Admin/Sell/Product/EventListener/CombinationListener.php:60]: No be cause when you call `get` on a form with a non-existing field an exception is thrown That's why we need to use `has` Besides even when get returns an elements it returns the form field not the data it contains.

So here I need to test:
- with `has` I check that the form field is indeed present (it can be removed by a configuration)
- if it is present but the data itself is empty I remove it since there is nothing to display (this field is used for display not edition)

File: src/Adapter/Product/Stock/Repository/StockMovementRepository.php (hunk 14 above)
Comment (jolelievre) [reply src/Adapter/Product/Stock/Repository/StockMovementRepository.php:65]: Why 3, why not? 😅. Actually this function is currently only used via the query `GetEmployeesStockMovements` and it contains a default limit with a const, so I think it's fine 🤷‍♂️

But this feature will be improved anyway in a later PR, I'm waiting for a more detailed specification by Mateus

**Comment by jolelievre** [comment]
@sowbiba actually the YAML linter is ok, everything is green n the CI

There are warnings present in the code review tab, but these warnings are present on ALL our PRs currently, it is related to our configuration which needs to be improved globally, nothing related to this PR.

**Review by NeOMakinG — APPROVED** [review]

**Comment by florine2623** [comment]
Hello @jolelievre ,

Tested again on Standard Product, works like a charm !
It is QA approved, thanks !

**Comment by Progi1984** [comment]
Thanks @jolelievre & @florine2623
